### PR TITLE
[SCB-607] Accesslog support invocation context item

### PR DESCRIPTION
**Pull request by yhs0092**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SCB) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SCB-XXX] Fixes bug in ApproximateQuantiles`, where you replace `SCB-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
See details in [SCB-607](https://issues.apache.org/jira/browse/SCB-607)

- set Invocation into RoutingContext
- add an AccessLogItem to get content from invocation context

### CODE REVIEW

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 1]
```diff
@@ -249,4 +265,21 @@ protected void onExecuteHttpServerFiltersFinish(Response response, Throwable e)
...
+  }
+
+  /**
+   * After the {@link AbstractRestInvocation#invocation} is instantiated, the {@link AbstractRestInvocation#invocation}
+   * will be passed into {@link AfterCreateInvocationHandler#handle(Invocation)} so that users can do something with it.
+   */
+  public static interface AfterCreateInvocationHandler {
+    /**
+     * @see AfterCreateInvocationHandler
+     * @param invocation {@link AbstractRestInvocation#invocation}
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: This document refers itself, and can be deleted

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java [hunk 2]
```diff
@@ -249,4 +265,21 @@ protected void onExecuteHttpServerFiltersFinish(Response response, Throwable e)
       invocation.onFinish(response);
     }
   }
+
+  public void setAfterCreateInvocationHandler(
```
Comment (liubao68) [inline common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: Maybe create a new protected method is better.
protected void afterCreateInvocation()
protected void beforeCreateInvocation()
protected void createInvocation()

 And we can implement it in RestProducerInvocation, e.g
RestProducerInvocation invo = new RestProducerInvocation.java(RoutingContext context)
..

File: edge/edge-core/src/main/java/org/apache/servicecomb/edge/core/EdgeInvocation.java [hunk 3]
```diff
@@ -59,6 +59,7 @@ public void init(String microserviceName, RoutingContext context, String path,
     this.responseEx = new VertxServerResponseToHttpServletResponse(context.response());
     this.httpServerFilters = httpServerFilters;
     requestEx.setAttribute(RestConst.REST_REQUEST, requestEx);
+    setAfterCreateInvocationHandler(invocation -> context.put(RestConst.REST_INVOCATION_CONTEXT, invocation));
```
Comment (wujimin) [inline edge/edge-core/src/main/java/org/apache/servicecomb/edge/core/EdgeInvocation.java]: if just need to do something after createInvocation
just override and call super first is enough?

File: transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/VertxRestDispatcher.java [hunk 4]
```diff
@@ -182,6 +182,8 @@ private void onRequest(RoutingContext context) {
 
     RestProducerInvocation restProducerInvocation = new RestProducerInvocation();
     context.put(RestConst.REST_PRODUCER_INVOCATION, restProducerInvocation);
+    restProducerInvocation.setAfterCreateInvocationHandler(
```
Comment (wujimin) [inline transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/VertxRestDispatcher.java]: both edge and normal need to set this, so set it to be default action?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java (hunk 2 above)
Comment (yhs0092) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: I guess you mean to replace the `AfterCreateInvocationHandler` with the method `protected void afterCreateInvocation()` ?
But currently we cannot get RoutingContext directly in `AbstractRestInvocation`, unless we cast the type of  `AbstractRestInvocation.requestEx` and add a getter into `VertxServerRequestToHttpServletRequest` like below:
```java
protected void afterCreateInvocation() {
  if (!VertxServerRequestToHttpServletRequest.class.isAssignableFrom(this.requestEx.getClass())) {
    return;
  }
  VertxServerRequestToHttpServletRequest request = (VertxServerRequestToHttpServletRequest) this.requestEx;
  RoutingContext routingContext = request.getContext();
  routingContext.put(RestConst.REST_INVOCATION_CONTEXT, this.invocation);
}
```

Am I right?

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java (hunk 1 above)
Comment (yhs0092) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: `AfterCreateInvocationHandler` has been removed.

File: common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java (hunk 2 above)
Comment (yhs0092) [reply common/common-rest/src/main/java/org/apache/servicecomb/common/rest/AbstractRestInvocation.java]: The code has been refactored, please review the new commit

File: edge/edge-core/src/main/java/org/apache/servicecomb/edge/core/EdgeInvocation.java (hunk 3 above)
Comment (yhs0092) [reply edge/edge-core/src/main/java/org/apache/servicecomb/edge/core/EdgeInvocation.java]: New setting logic is in the overridden `createInvocation()` method, please review the new commit

File: transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/VertxRestDispatcher.java (hunk 4 above)
Comment (yhs0092) [reply transports/transport-rest/transport-rest-vertx/src/main/java/org/apache/servicecomb/transport/rest/vertx/VertxRestDispatcher.java]: The code has been refactored, please review the new commit

**Pull request merged**